src/shell/seats.rs

```diff
@@ -229,13 +245,31 @@ impl SeatExt for Seat<State> {
         self.user_data().get::<SeatId>().unwrap().0
     }
 
+    /// Returns the output that contains the cursor associated with a seat. Note that the window which has keyboard focus
+    /// may be on a different output. Currently, to get the focused output, first get the keyboard focus target and pass
+    /// it to get_focused_output in the shell.
     fn active_output(&self) -> Output {
         self.user_data()
             .get::<ActiveOutput>()
             .map(|x| x.0.lock().unwrap().clone())
             .unwrap()
     }
 
+    /// Returns the output which currently has keyboard focus. If no window has keyboard focus (e.g. when there are no windows)
+    /// the focused output will be the same as the active output.
+    fn focused_output(&self) -> Option<Output> {
+        if self.get_keyboard().is_some() {
+            Some(
+                self.user_data()
+                    .get::<FocusedOutput>()
+                    .map(|x| x.0.lock().unwrap().clone())
+                    .unwrap(),
+            )
+        } else {
+            None
+        }
+    }
```

Comment:
I wonder, if we really want to have this function or just always query `Shell` with the seat instead. My issue is, that this is storing the same state in multiple ways. If we ever forget to update `focused_output` when changing focus, this will break behaviour quite drastically.

Reply:
It's up to you, though the places where this is being set is the same as when keyboard focus target is updated. 

In the meantime, I've reverted the changes to `input/mod.rs` on my end and will force push once I implement changes from the other reviews.

Reply:
we could wrap all 3 actions (set Active focus, set focused output, set keyboard focus) inside a new private function that is called internally by `setfocus` and `refresh_focus` to make sure that those actions always happen together. That would probably be the best way to implement focus follows pointer if set anyway

Reply:
> we could wrap all 3 actions (set Active focus, set focused output, set keyboard focus) inside a new private function that is called internally by `setfocus` and `refresh_focus` to make sure that those actions always happen together. That would probably be the best way to implement focus follows pointer if set anyway

Sounds good to me.

Reply:
I'll mark this as resolved since I've moved setting these to an internal function